module/zfs/zap.c

```diff
@@ -850,10 +854,29 @@ fzap_add_cd(zap_name_t *zn,
 	if (err == 0) {
 		zap_increment_num_entries(zap, 1, tx);
 	} else if (err == EAGAIN) {
+		/*
+		 * If the last two expansions did not help, there is no point
+		 * trying to expand again
+		 */
+		if (expand_retries > MAX_EXPAND_RETRIES && prev_l == l) {
+			err = ENOSPC;
```

Comment:
You should use `err = SET_ERROR(ENOSPC);` here, similar to what's done a few lines up for `EEXIST`.

Reply:
Done